packages/fetch-error-handler/lib/create-handler.js

```diff
@@ -0,0 +1,141 @@
+const {
+	HttpError,
+	UpstreamServiceError
+} = require('@dotcom-reliability-kit/errors');
+
+/**
+ * @typedef {object} ErrorHandlerOptions
+ * @property {string} [upstreamSystemCode]
+ *     The system code of the upstream system that the `fetch` makes a request to.
+ */
+
+/**
+ * @typedef {object} FetchResponse
+ * @property {boolean} ok
+ *     Whether the fetch was successful.
+ * @property {number} status
+ *     The response HTTP status code.
+ * @property {string} url
+ *     The URL of the response.
+ */
+
+/* eslint-disable jsdoc/valid-types */
+// The ESLint JSDoc plugin we're using marks this as invalid despite it being fine.
+// This is covered by the following GitHub issues. If these are closed then this should
+// no longer be an issue:
+// https://github.com/gajus/eslint-plugin-jsdoc/issues/145
+// https://github.com/jsdoc-type-pratt-parser/jsdoc-type-pratt-parser/issues/96
+// https://github.com/jsdoctypeparser/jsdoctypeparser/issues/50
+/**
+ * @typedef {<Response extends FetchResponse>(fetchPromise: Response | Promise<Response>) => Promise<Response>} FetchErrorHandler
+ */
+/* eslint-enable jsdoc/valid-types */
+
+/**
+ * Create an fetch error handler function.
+ *
+ * @param {ErrorHandlerOptions} [options]
+ *     Configuration options for the handler.
+ * @returns {FetchErrorHandler}
+ *     Returns an error handler for use with a `fetch` function.
+ */
+function createFetchErrorHandler(options = {}) {
+	const { upstreamSystemCode } = options;
+	const relatesToSystems =
+		typeof upstreamSystemCode === 'string' ? [upstreamSystemCode] : [];
+
+	return async function handleFetchErrors(input) {
+		let response = input;
+
+		// If input is a promise, resolve it
+		if (input instanceof Promise) {
+			response = await input;
+		}
+
+		// Check whether the value we were given is a valid response object
+		if (!isFetchResponse(response)) {
+			// This is not an operational error because the invalid
+			// input is highly likely to be a programmer error
+			throw Object.assign(
+				new TypeError(
+					'Fetch handler must be called with a `fetch` response object or a `fetch` promise'
+				),
+				{ code: 'FETCH_ERROR_HANDLER_INVALID_INPUT' }
+			);
+		}
+
+		// If the response isn't OK, we start throwing errors
+		if (!response.ok) {
+			// Parse the response URL so we can use the hostname in error messages
+			let responseHostName = 'unknown';
+			if (typeof response.url === 'string') {
+				try {
+					const url = new URL(response.url);
+					responseHostName = url.hostname;
+				} catch (_) {}
```

Comment:
curious: my understanding is that if we want to communicate to other devs that we've intentionally put a 'dummy' placeholder, then we use an underscore. Is this why the underscore is passed into the catch parameter? and if so, why do we have the catch and not just take the code out of the try/catch block? 🙇🏽

Reply:
Thank you as always 😸  lovely explanation and this makes sense. Yeah I think the comment above would be helpful to help clarify